ldap_auth_provider.py

```diff
@@ -404,6 +404,74 @@ class _LdapConfig(object):
 
         return ldap_config
 
+    def _get_server(self, get_info: Optional[str] = None) -> ldap3.ServerPool:
+        """Constructs ServerPool from configured LDAP URIs
+
+        Args:
+            get_info (str, optional): specifies if the server schema and server
+            specific info must be read. Defaults to None.
+
+        Returns:
+            [ServerPool]: servers groupped to ServerPool
+        """
+        return ldap3.ServerPool(
+            [
+                ldap3.Server(
+                    uri,
+                    get_info=get_info,
+                    tls=ldap3.Tls(validate=ssl.CERT_REQUIRED)
+                )
+                for uri in self.ldap_uris
+            ],
+        )
+
+    async def _fetch_root_domain(self) -> str:
+        """Fetches root domain from LDAP and saves it to ``self.ldap_root_domain``
+
+        Returns:
+            str: The root domain of Active Directory forest
+        """
+        if self.ldap_root_domain is not False:
+            return self.ldap_root_domain
+
+        if self.ldap_mode != LDAPMode.SEARCH:
+            logger.warning("Fetching root domain is supported in search mode only")
+            self.ldap_root_domain = None
+            return self.ldap_root_domain
+
+        server = self._get_server(get_info=ldap3.DSA)
+        result, conn = await self._ldap_simple_bind(
+            server=server,
+            bind_dn=self.ldap_bind_dn,
+            password=self.ldap_bind_password,
+        )
+
+        if result:
+            self.ldap_root_domain = None
+
+            if (
+                conn.server.info.other
+                and "rootDomainNamingContext" in conn.server.info.other
+                and conn.server.info.other["rootDomainNamingContext"]
+            ):
+                # conn.server.info.other["rootDomainNamingContext"][0]
+                # is of the form DC=example,DC=org
+                self.ldap_root_domain = ".".join(
+                    [
+                        dc.split("=")[1] for dc
+                        in conn.server.info.other["rootDomainNamingContext"][0].split(",")
+                        if "=" in dc
+                    ]
+                )
+                logger.info('Obtained root domain "%s"', self.ldap_root_domain)
+        else:
+            logger.warning("Unable to get root domain")
```

Comment:
the comment further up says that `self.ldap_root_domain` will be set to None if there was an error fetching the root domain, but we seem to leave it set to `False` here. please could you make sure the comments match the implementation?